### PR TITLE
resource/aws_route53_resolver_firewall_domain_list: Add new resource

### PR DESCRIPTION
**Pull request by gazoakley**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #18520

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSRoute53ResolverFirewallDomainList'        
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSRoute53ResolverFirewallDomainList -timeout 180m
=== RUN   TestAccAWSRoute53ResolverFirewallDomainList_basic
=== PAUSE TestAccAWSRoute53ResolverFirewallDomainList_basic
=== RUN   TestAccAWSRoute53ResolverFirewallDomainList_domains
=== PAUSE TestAccAWSRoute53ResolverFirewallDomainList_domains
=== RUN   TestAccAWSRoute53ResolverFirewallDomainList_disappears
=== PAUSE TestAccAWSRoute53ResolverFirewallDomainList_disappears
=== RUN   TestAccAWSRoute53ResolverFirewallDomainList_tags
=== PAUSE TestAccAWSRoute53ResolverFirewallDomainList_tags
=== CONT  TestAccAWSRoute53ResolverFirewallDomainList_basic
=== CONT  TestAccAWSRoute53ResolverFirewallDomainList_tags
=== CONT  TestAccAWSRoute53ResolverFirewallDomainList_disappears
=== CONT  TestAccAWSRoute53ResolverFirewallDomainList_domains
--- PASS: TestAccAWSRoute53ResolverFirewallDomainList_disappears (88.79s)
--- PASS: TestAccAWSRoute53ResolverFirewallDomainList_basic (107.95s)
--- PASS: TestAccAWSRoute53ResolverFirewallDomainList_tags (137.11s)
--- PASS: TestAccAWSRoute53ResolverFirewallDomainList_domains (348.92s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	352.634s
```

### CODE REVIEW

**Comment by gazoakley** [comment]
This currently branches from #18544 - if a later build deps bump gets merged this can be rebased

File: aws/resource_aws_route53_resolver_firewall_domain_list.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+		if err != nil {
+			return fmt.Errorf("error waiting for Route 53 Resolver DNS Firewall domain list (%s) domains to be updated: %w", d.Id(), err)
+		}
+	}
+
+	if d.HasChange("tags") {
+		o, n := d.GetChange("tags")
+		if err := keyvaluetags.Route53resolverUpdateTags(conn, d.Get("arn").(string), o, n); err != nil {
+			return fmt.Errorf("error updating Route53 Resolver DNS Firewall domain list (%s) tags: %s", d.Get("arn").(string), err)
```
Comment (ewbankkit) [inline aws/resource_aws_route53_resolver_firewall_domain_list.go:160]:
```suggestion
			return fmt.Errorf("error updating Route53 Resolver DNS Firewall domain list (%s) tags: %w", d.Get("arn").(string), err)
```

File: aws/resource_aws_route53_resolver_firewall_domain_list.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+		return fmt.Errorf("error setting tags: %w", err)
+	}
+
+	return nil
+}
+
+func resourceAwsRoute53ResolverFirewallDomainListUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).route53resolverconn
+
+	if v, ok := d.GetOk("domains"); ok && d.HasChange("domains") {
```
Comment (ewbankkit) [inline aws/resource_aws_route53_resolver_firewall_domain_list.go:139]: Can all the domains for a list be deleted? If so will need to add a call to `UpdateFirewallDomains` with `Operation = REMOVE`.

**Comment by gazoakley** [comment]
Hi @ewbankkit - PR updated and conflicts with main resolved

File: aws/resource_aws_route53_resolver_firewall_domain_list.go (hunk 2 above)
Comment (gazoakley) [reply aws/resource_aws_route53_resolver_firewall_domain_list.go]: Updated - I'm now calling GetChanges to remove the old entries if there aren't any new entries anymore.

File: aws/resource_aws_route53_resolver_firewall_domain_list.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+		Schema: map[string]*schema.Schema{
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
```
Comment (ewbankkit) [inline aws/resource_aws_route53_resolver_firewall_domain_list.go:35]: No need for explicit `id` attribute.

File: aws/resource_aws_route53_resolver_firewall_domain_list.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+	if firewallDomainList == nil {
+		log.Printf("[WARN] Route 53 Resolver DNS Firewall domain list (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	arn := aws.StringValue(firewallDomainList.Arn)
+	d.Set("arn", arn)
+	d.Set("id", firewallDomainList.Id)
```
Comment (ewbankkit) [inline aws/resource_aws_route53_resolver_firewall_domain_list.go:104]: Remove.

**Comment by gazoakley** [comment]
Many thanks for the review @ewbankkit - updated with feedback 👍

**Review by ewbankkit — APPROVED** [review]
LGTM.

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSRoute53ResolverFirewallDomainList_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSRoute53ResolverFirewallDomainList_ -timeout 180m
=== RUN   TestAccAWSRoute53ResolverFirewallDomainList_basic
=== PAUSE TestAccAWSRoute53ResolverFirewallDomainList_basic
=== RUN   TestAccAWSRoute53ResolverFirewallDomainList_domains
=== PAUSE TestAccAWSRoute53ResolverFirewallDomainList_domains
=== RUN   TestAccAWSRoute53ResolverFirewallDomainList_disappears
=== PAUSE TestAccAWSRoute53ResolverFirewallDomainList_disappears
=== RUN   TestAccAWSRoute53ResolverFirewallDomainList_tags
=== PAUSE TestAccAWSRoute53ResolverFirewallDomainList_tags
=== CONT  TestAccAWSRoute53ResolverFirewallDomainList_basic
=== CONT  TestAccAWSRoute53ResolverFirewallDomainList_tags
=== CONT  TestAccAWSRoute53ResolverFirewallDomainList_disappears
=== CONT  TestAccAWSRoute53ResolverFirewallDomainList_domains
--- PASS: TestAccAWSRoute53ResolverFirewallDomainList_disappears (83.47s)
--- PASS: TestAccAWSRoute53ResolverFirewallDomainList_basic (100.75s)
--- PASS: TestAccAWSRoute53ResolverFirewallDomainList_tags (114.65s)
--- PASS: TestAccAWSRoute53ResolverFirewallDomainList_domains (334.35s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	334.458s
```

**Comment by breathingdust** [comment]
LGTM :rocket: Thanks @gazoakley!

Verified Acceptance Tests in Commercial (us-west-2)

```console
make testacc TESTARGS='-run=TestAccAWSRoute53ResolverFirewallDomainList_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSRoute53ResolverFirewallDomainList_ -timeout 180m
=== RUN   TestAccAWSRoute53ResolverFirewallDomainList_basic
=== PAUSE TestAccAWSRoute53ResolverFirewallDomainList_basic
=== RUN   TestAccAWSRoute53ResolverFirewallDomainList_domains
=== PAUSE TestAccAWSRoute53ResolverFirewallDomainList_domains
=== RUN   TestAccAWSRoute53ResolverFirewallDomainList_disappears
=== PAUSE TestAccAWSRoute53ResolverFirewallDomainList_disappears
=== RUN   TestAccAWSRoute53ResolverFirewallDomainList_tags
=== PAUSE TestAccAWSRoute53ResolverFirewallDomainList_tags
=== CONT  TestAccAWSRoute53ResolverFirewallDomainList_basic
=== CONT  TestAccAWSRoute53ResolverFirewallDomainList_tags
=== CONT  TestAccAWSRoute53ResolverFirewallDomainList_disappears
=== CONT  TestAccAWSRoute53ResolverFirewallDomainList_domains
--- PASS: TestAccAWSRoute53ResolverFirewallDomainList_disappears (87.72s)
--- PASS: TestAccAWSRoute53ResolverFirewallDomainList_basic (91.63s)
--- PASS: TestAccAWSRoute53ResolverFirewallDomainList_tags (114.16s)
--- PASS: TestAccAWSRoute53ResolverFirewallDomainList_domains (327.32s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	330.205s
```

Verified Acceptance Tests in GovCloud (us-gov-west-1)

```console
make testacc TESTARGS='-run=TestAccAWSRoute53ResolverFirewallDomainList_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSRoute53ResolverFirewallDomainList_ -timeout 180m
=== RUN   TestAccAWSRoute53ResolverFirewallDomainList_basic
=== PAUSE TestAccAWSRoute53ResolverFirewallDomainList_basic
=== RUN   TestAccAWSRoute53ResolverFirewallDomainList_domains
=== PAUSE TestAccAWSRoute53ResolverFirewallDomainList_domains
=== RUN   TestAccAWSRoute53ResolverFirewallDomainList_disappears
=== PAUSE TestAccAWSRoute53ResolverFirewallDomainList_disappears
=== RUN   TestAccAWSRoute53ResolverFirewallDomainList_tags
=== PAUSE TestAccAWSRoute53ResolverFirewallDomainList_tags
=== CONT  TestAccAWSRoute53ResolverFirewallDomainList_basic
=== CONT  TestAccAWSRoute53ResolverFirewallDomainList_tags
=== CONT  TestAccAWSRoute53ResolverFirewallDomainList_disappears
=== CONT  TestAccAWSRoute53ResolverFirewallDomainList_domains
--- PASS: TestAccAWSRoute53ResolverFirewallDomainList_disappears (85.58s)
--- PASS: TestAccAWSRoute53ResolverFirewallDomainList_basic (89.85s)
--- PASS: TestAccAWSRoute53ResolverFirewallDomainList_tags (110.42s)
--- PASS: TestAccAWSRoute53ResolverFirewallDomainList_domains (344.87s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	347.720s
```

**Pull request merged**